Forms system

Defective4 opened the pull request:
This pull request adds the forms system to the bot.
The forms can be attached to any of the bot's messages. Users can fill in a form by clicking on the button added to the message the form is attached. After doing so, the user is presented with a modal dialog containing up to 5 fields defined in the form.
Upon submission, the contents of the fields are sent to a dedicated channel (configurable per form).

Included in this PR are commands to create, manage, and customize forms.
Forms can also be configured to accept only one submission per user. This was achieved by keeping track of who had submitted the form before in the dedicated forms repository.

Below is a video showcasing how the system works, along with most of the commands:
[~~link~~](https://services.defective.is-a.dev/files/api/public/dl/jgc5P06d?inline=true) (the link is broken by now)

I'm aware this can be improved in some areas. I'm looking forward to your reviews!

danthe1st commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/model/FormField.java:
```diff
@@ -0,0 +1,95 @@
+import net.dv8tion.jda.api.interactions.components.text.TextInputStyle;
+
+/**
+ * Represents a form field.
```
Maybe add a short descrion of what a form field is.

danthe1st commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/model/FormField.java:
```diff
@@ -0,0 +1,95 @@
+	private final int min;
+	private final String placeholder;
+	private final boolean required;
+	private final String style;
```
It would also be possible to use the `TextInputStyle` annotation here and then use methods like `.name()`/`valueOf()`

danthe1st commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/model/FormUser.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Represents an user who submitted a form.
+ */
+public class FormUser {
```
This could be a record as well (if it is even needed)

Defective4 commented on src/main/resources/database/migrations/09-08-2025_forms.sql:
```diff
@@ -0,0 +1,23 @@
+CREATE TABLE forms (
+    form_id BIGINT NOT NULL,
+    form_data VARCHAR NOT NULL,
```
I'm not that experienced when it comes to dealing with database stuff, and I NEVER worked with H2 databases - I wasn't even aware there was a JSON datatype

danthe1st commented on src/main/resources/database/migrations/09-08-2025_forms.sql:
```diff
@@ -0,0 +1,23 @@
+    form_id BIGINT NOT NULL,
+    form_data VARCHAR NOT NULL,
+    title VARCHAR NOT NULL,
+    submit_message VARCHAR,
```
¯\\\_(ツ)_/¯

Defective4 commented on src/main/resources/database/migrations/09-08-2025_forms.sql:
```diff
@@ -0,0 +1,23 @@
+    user_id VARCHAR NOT NULL,
+    form_id BIGINT NOT NULL,
+    user_name VARCHAR NOT NULL,
+    PRIMARY KEY ("timestamp")
```
Yeah, that's a great idea

danthe1st commented on src/main/resources/database/migrations/09-08-2025_forms.sql:
```diff
@@ -0,0 +1,23 @@
+    "timestamp" BIGINT NOT NULL,
+    user_id VARCHAR NOT NULL,
+    form_id BIGINT NOT NULL,
+    user_name VARCHAR NOT NULL,
```
> Unless the bot has member cache enabled

Member caching is disabled but `Guild#retrieveMembersByIds` (and similar) exist.

Defective4 replied:
This method is limited to 100 members at a time and can time out. I know it shouldn't be an issue in most cases, but I'd rather not query Discord's API when retrieving the list of submissions, so for the time being, I'll leave it as-is if you don't mind

danthe1st reviewed (changes requested):
Round 2 of reviewing, `FormInteractionManager`, `/form add-field`, `/form attach`

Defective4 commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/commands/AddFieldFormSubcommand.java:
```diff
@@ -0,0 +1,103 @@
+	@Override
+	public void execute(SlashCommandInteractionEvent event) {
+
+		event.deferReply(true).queue();
```
There is no particular reason; it was just convenient for me not to do `setEphemeral` on every response. Plus, there are database operations in every command, and I'm used to deferring the reply when any kind of I/O is involved.

Defective4 commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/FormInteractionManager.java:
```diff
@@ -0,0 +1,275 @@
+	private static MessageEmbed createSubmissionEmbed(FormData form, List<ModalMapping> values, Member author) {
+		EmbedBuilder builder = new EmbedBuilder().setTitle("New form submission received")
+				.setAuthor(author.getEffectiveName(), null, author.getEffectiveAvatarUrl()).setTimestamp(Instant.now());
+		builder.addField("Sender", author.getAsMention(), true).addField("Title", form.getTitle(), true);
```
You mean the author of the form?

danthe1st replied:
I meant the person who submitted the form.

Defective4 commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/FormInteractionManager.java:
```diff
@@ -0,0 +1,275 @@
+				.setAuthor(author.getEffectiveName(), null, author.getEffectiveAvatarUrl()).setTimestamp(Instant.now());
+		builder.addField("Sender", author.getAsMention(), true).addField("Title", form.getTitle(), true);
+
+		int len = Math.min(values.size(), form.getFields().size());
```
Close, they will be empty. It's actually briefly shown in my recording, although I probably should add a clear indicator that the user omitted the field.

Defective4 commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/FormInteractionManager.java:
```diff
@@ -0,0 +1,275 @@
+		for (int i = 0; i < len; i++) {
+			ModalMapping mapping = values.get(i);
+			FormField field = form.getFields().get(i);
+			String value = mapping.getAsString();
```
That's a good point, I'll definitely do that

danthe1st commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/model/FormData.java:
```diff
@@ -70,10 +71,14 @@ public FormData(long id, List<FormField> fields, String title, String submitChan
-	 * @return Lsit of layout components for use in the submission modal.
+	 * @return List of layout components for use in the submission modal.
 	 */
 	public LayoutComponent[] createComponents() {
```
I don't think this needs to be an array.

danthe1st commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/model/FormField.java:
```diff
@@ -44,46 +16,8 @@ public FormField(String label, int max, int min, String placeholder, boolean req
+				.setRequired(required()).setValue(value()).build();
 	}
 
 	@Override
```
Any reason you are still overriding the `toString` here?

danthe1st commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/model/FormData.java:
```diff
@@ -70,10 +71,14 @@ public FormData(long id, List<FormField> fields, String title, String submitChan
 		this.onetime = onetime;
 	}
 
+	public boolean isAttached() {
```
You are often using the check of calling `isAttached()` and then calling `Optional#get` if `isAttached` returned `true`. Alternatively, you could create another record for the channel+message and return an `Optional` of that record. Then you could use `Optional`s capabilities instead of `Optional#get` (and tbh I'd prefer `Optional#orElseThrow` over `Optional#get` because that's more clear about what it does):
```java
form.getMessageInformation().ifPresentOrElse(messageInfo -> doSomething(messageInfo), () -> notAttached()); // or similar
```
or
```
String text = form.getMessageInformation().map(messageInfo ->
        String.format("[Link](https://discord.com/channels/%s/%s/%s)",
        guild.getId(),
        messageInfo.channelId(),
        messageInfo.messageId()),
    ).orElse(() -> "*Not attached*");

danthe1st commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/commands/SubmissionsDeleteFormSubcommand.java:
```diff
@@ -29,11 +30,9 @@ public class SubmissionsDeleteFormSubcommand extends Subcommand implements AutoC
-								true),
-						new OptionData(OptionType.STRING, "user-id", "User to delete submissions of", true, true)));
+		setCommandData(new SubcommandData("submissions-delete", "Deletes submissions of a user in the form")
+				.addOptions(new OptionData(OptionType.INTEGER, "form-id", "The ID of a form to get submissions for",
```
_get_ submissions for is misleading.

Defective4 replied:
Yup, it's just another oversight of mine. I did a lot of copy-pasting when creating the commands, and it seems I forgot to change the parameter description here

danthe1st reviewed (changes requested):
Please still act on the other requested changes like unifying autocomplete.

Defective4 commented on src/main/resources/database/migrations/09-08-2025_forms.sql:
```diff
@@ -1,23 +1,38 @@
+
+CREATE TABLE FORM_SUBMISSIONS (
+    ID BIGINT NOT NULL AUTO_INCREMENT,
+    MESSAGE_ID BIGINT NOT NULL,
```
Right, I guess not

Defective4 commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/commands/DeleteFormSubcommand.java:
```diff
@@ -48,8 +48,9 @@ public void execute(SlashCommandInteractionEvent event) {
-		if (form.getMessageChannel() != null && form.getMessageId() != null) {
+		if (form.isAttached()) {
 			DetachFormSubcommand.detachFromMessage(form, event.getGuild());
+			// TODO send a warning
```
Yup, that's the plan.

Defective4 commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/dao/FormsRepository.java:
```diff
@@ -133,7 +111,7 @@ public void detachForm(FormData form) {
 	 * @return A list of forms
 	 */
 	public List<FormData> getAllForms() {
```
That's true. I made this method in case I needed a list of all forms somewhere else.

Defective4 commented:
> By the way, the [last build](https://github.com/Java-Discord/JavaBot/actions/runs/17801898938/job/50603681493) failed because of checkstyle.
I know. I will take care of checkstyle once I'm done writing the actual code.

Defective4 commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/model/FormData.java:
```diff
@@ -0,0 +1,161 @@
+	public FormData(long id, List<FormField> fields, String title, String submitChannel, String submitMessage,
+			String messageId, String messageChannel, long expiration, boolean closed, boolean onetime) {
+		this.id = id;
+		this.fields = Objects.requireNonNull(fields);
```
Is there any reason I should use `List#copyOf` instead of `Collections#unmodifiableList`?

danthe1st replied:
`Collections.unmodifiableList` returns a _view_ on the `List` i.e. if the original `List` is modified, these changes are propagated to the unmodifiable view which can have unintended side affects.
Example:
```java
List<String> elems = new ArrayList<>();
elems.add("Hello");
List<String> thisIsNotImmutable = Collections.unmodifiableList(elems);
System.out.println(thisIsNotImmutable);
elems.add("World");
System.out.println(thisIsNotImmutable);
```

Defective4 commented:
Also, please keep in mind that even though I'm pushing new commits, I'm still not done addressing most of the issues. I have a habit of pushing immediately after committing new changes to the code.

Defective4 commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/commands/ModifyFormSubcommand.java:
```diff
@@ -1,78 +1,104 @@
-				OptionMapping::getAsString);
-		long expiration;
+		String title = event.getOption("title", oldForm.title(), OptionMapping::getAsString);
+		long submitChannel = event.getOption("submit-channel", oldForm.submitChannel(), OptionMapping::getAsLong);
```
Eclipse didn't replace those, oops

danthe1st commented on src/main/java/net/discordjug/javabot/systems/staff_commands/forms/dao/FormsRepository.java:
```diff
@@ -171,12 +149,12 @@ public Map<FormUser, Integer> getAllSubmissions(FormData form) {
 	 *
 	 * @param formId form ID to query
-	 * @return optional containing the form, or empty if the form was not found.
+	 * @return optional form
```
@Defective4 I think this is still missing.
